### PR TITLE
support response body stream

### PR DESCRIPTION
**Pull request by Anthony-Dong**
why:
When developing HTTP proxy tools, we encountered the need to forward large packets, and the second is to support streaming forwarding, so we need to modify the sdk.

For security reasons,`Response.BodyStream()` function return `io.ReadCloser` and lets the user manually release the resource . 

Finally, I want request to support stream and open it up to developers.

### CODE REVIEW

File: streaming.go [hunk 1]
```diff
@@ -10,7 +10,10 @@ import (
 )
 
 type requestStream struct {
-	header          *RequestHeader
+	header interface {
+		ContentLength() int
+		ReadTrailer(r *bufio.Reader) error
+	}
```
Comment (erikdubbelboer) [inline streaming.go:16]: Probably better to define an interface type for this as it's used twice in this file.

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]
Looks good at a first glance, I still have to take a deep look at it.

File: streaming.go (hunk 1 above)
Comment (Anthony-Dong) [reply streaming.go]: Thank you. I have modified it.

**Comment by byene0923** [comment]
i'm also look deep into it, i think the code can be more elegant, do you think use `Requeststream`   in response is a good way?

**Comment by Anthony-Dong** [comment]
> 我也在深入研究它，我认为代码可以更优雅，你认为使用`Requeststream` 响应是一个好方法吗？



> i'm also look deep into it, i think the code can be more elegant, do you think use `Requeststream` in response is a good way?

**Comment by Anthony-Dong** [comment]
> i'm also look deep into it, i think the code can be more elegant, do you think use `Requeststream` in response is a good way?



> 

I think it's just a naming issue, since request stream was already supported.

**Comment by byene0923** [comment]
> > i'm also look deep into it, i think the code can be more elegant, do you think use `Requeststream` in response is a good way?
> 
> > 
> 
> I think it's just a naming issue, since request stream was already supported.

you could change the name  such as `bodyStream`

**Comment by Anthony-Dong** [comment]
@erikdubbelboer do you have any questions ?

File: http.go [hunk 2]
```diff
@@ -293,6 +297,40 @@ func (resp *Response) BodyWriter() io.Writer {
...
 
+// BodyStream returns io.Reader
+//
+// You must close it after you use it.
+func (resp *Response) BodyStream() io.ReadCloser {
+	if resp.bodyStream == nil {
+		resp.bodyStream = bytes.NewReader(resp.Body())
+	}
+	return newCloseReader(resp.bodyStream, resp.closeBodyStream)
+}
```
Comment (erikdubbelboer) [inline http.go:308]: Is this method useful? We could also not add it for now.

File: http.go [hunk 3]
```diff
@@ -92,6 +92,10 @@ type Response struct {
 	// Relevant for bodyStream only.
 	ImmediateHeaderFlush bool
 
+	// StreamBody enables response body streaming.
+	// Response.BodyStream() get response body.
```
Comment (erikdubbelboer) [inline http.go:96]:
```suggestion
	// StreamBody enables response body streaming.
	// Use SetBodyStream to set the body stream.
```

File: http.go [hunk 4]
```diff
@@ -293,6 +297,40 @@ func (resp *Response) BodyWriter() io.Writer {
...
+	return &closeReader{Reader: r, closeFunc: closeFunc}
+}
+
+func (c *closeReader) Close() error {
+	if c.closeFunc == nil {
+		return nil
+	}
+	c.closeOnce.Do(func() {
+		c.err = c.closeFunc()
+	})
```
Comment (erikdubbelboer) [inline http.go:330]: Why is this `sync.Once` needed here? If `Close` can be called multiple times, shouldn't `closeFunc` then also not be able to be called multiple times?

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

File: http.go (hunk 4 above)
Comment (Anthony-Dong) [reply http.go:330]: if SetBodyStream is true, the clientConn does not immediately close.  need to close  it manually.  close clientConn does not need to be executed multiple times.

File: http.go (hunk 2 above)
Comment (Anthony-Dong) [reply http.go:308]: We hope to read the response body through stream to prevent excessive memory pressure

File: http.go (hunk 4 above)
Comment (Anthony-Dong) [reply http.go:330]: For example, `sync.Pool` has the same object and `clientConn` is closed multiple times

**Comment by Anthony-Dong** [comment]
> 



> > > i'm also look deep into it, i think the code can be more elegant, do you think use `Requeststream` in response is a good way?
> > 
> > 
> > > 
> > 
> > 
> > I think it's just a naming issue, since request stream was already supported.
> 
> you could change the name such as `bodyStream`

`type responseStream requestStream` .  might be better

File: http.go (hunk 4 above)
Comment (erikdubbelboer) [reply http.go:330]: I'm not sure I understand. Can you show how `Close` would be called multiple times?

Comment (Anthony-Dong) [reply http.go:330]: eg: `stream.Close()` Execute multiple times

```go
resp := AcquireResponse()
resp.StreamBody = true
request := AcquireRequest()
request.SetRequestURI(server.URL + "?chunked=true")
if err := Do(request, resp); err != nil {
	t.Fatal(err)
}
stream := resp.BodyStream()
defer func() {
	// exec multiple times
	stream.Close()
	stream.Close()
	stream.Close()
}()
content, _ := io.ReadAll(stream)
if string(content) != "0123456789" {
	t.Fatalf("unexpected body content, got: %#v, want: %#v", string(content), "0123456789")
}
```

Comment (Anthony-Dong) [reply http.go]: I thought it wouldn't be good to return an io.ReadCloser directly, so I added the CloseBodyStream method @erikdubbelboer

Comment (erikdubbelboer) [reply http.go]: I did some investigating into this issue, https://github.com/valyala/fasthttp/issues/1504, and it seems it is cause by Close being called multiple times. So I understand the `sync.Once` now. But I'm wondering if we should fix the double close instead.

Comment (Anthony-Dong) [reply http.go]: `perIPConn#Close()` method is also wrapped in `sync.Once` ?  It also seems reasonable and easier to deal with

Comment (Anthony-Dong) [reply http.go]: I think it depends on whether there is concurrent execution of 'closeBodyStream'. If there is, the same problem will indeed occur, but this is an unreasonable use behavior and concurrency framework defect.  @erikdubbelboer

**Comment by erikdubbelboer** [comment]
Finally had the time to look, looks good. Thanks!

**Pull request merged**